CenterForOpenScience/osf.io
# [ENG-152] Normalize user tags.

## Purpose
This ticket helps identify how users come into the system. For this purpose, we plan to change **what** and **how** system tags are added to OSFUser instances. First, a bit of background:

Currently, there are two types of user system tags that tracks how user come into the system:
  - Campaign tags
    - These are the tags added to users when they register from a campaign landing page such as prereg, registered report and osf4m.
  - Provider tags
    - These are added to users when they register from a preprint provider landing page, including OSFPreprints.

However, there are two major drawbacks to how we utilize these system tags: the naming is ambiguous, and we don't add any tags to unregistered contributors coming into the system. And if an unreg contrib is added to multiple OSF nodes and preprints, we don't really know from which claim link the user confirms his/her account.

Therefore, this ticket implements a three-step plan to make sure we can identify how users come in, and in case of unregistered contributors, from which provider's preprints or campaign's nodes the user claim their accounts.

The first step is normalization, which is to name currently existing tags in a unified format. We call these tags "source tags" because they help identify the source of the users. We also add a separate type of tags called "claimed tags", which are used to identify from what campagin/provider the user "claims" his/her account. Source tags are added to all users, while claimed tags are only added to unregistered contributors upon claiming account.

The second step is backfilling source tags to unregistered contributors. Because the corresponding `unclaimed_records` entry is deleted when the user claims the account, we have no way but to look at node/preprint log entries to identify the node/preprint the user is added as an unreg contrib. For the same reason, we cannot accurately backfill claimed tags, so we decided not to.

The third step is making sure all incoming users are properly tagged. This includes adding source tags to users when they are created and adding claimed tags to unregistered contributors when they claim accounts.


## Changes

According to the plan, the following changes are made:

Step 1 - Normalization:
- Write a script to change names of existing provider source tags 
- Write a script to add provider claimed tags for existing provider source tags 
- Write a script to change names of existing campaign source tags 
- Write a script to add campaign claimed tags for existing campaign source tags 
- Write a script to add new provider tags 
  - `source:provider|osf` 
  - `claimed:provider|osf` 
- Write a script to add new campaign tags 
  - `source:campaign|prereg` 
    - Migrate user created before after 2019-01-01 but with `source:campaign|prereg_challenge` tag to new tag 
  - `claimed:campaign|prereg` 

Step 2 - Backfill:
- Write a script to backfill source tags to unreg contributors added to OSF4M nodes 
- Write a script to backfill source tags to unreg contributors added to preprints/supplemental node to preprints 
  - For supplemental nodes created less than 10 minutes before the creation of the preprint, we add provider source tags to unregistered contributors added to these nodes. 
  - For supplemental nodes created more than 10 minutes but less than a day before the creation of the preprint, we do nothing to its unreg contributors. 
  - For supplemental nodes created more than a day before the creation of the preprint, we add `source|provider|osf` to unregistered contributors added to these nodes. 
- Backfill	 `source:provider|osf` to users not invited but with no source tags. 
- No need to backfill source tags to unregistered contributors for prereg challenge/prereg/registered report because we don’t know what they are. 

Step 3 - Make sure we add tags to users on their way in.
- Add source tags to OSFUsers when they register through the landing pages or by email (in case of OSF4M). Specifically,
  - Add preprint provider source tags when they are registered from preprint landing pages.
  - Add registry provider source tags when they are registered from registry landing page.
  - Add prereg campaign source tag when they are registered on the prereg landing page.
  - Add registered report campaign source tag when they are registered on the RR landing page.
  - Add OSF4M campaign source tag when they are registered as a result of sending an email to meetings endpoint.
  - Otherwise, add osf provider source tag.
- Add campaign source tags to nodes when they are created from prereg/registered report landing pages. So that we can tell what nodes are created from such landing pages.
- Add source tags to unregistered contributors when they are added to a node/preprint. Specifically, 
  - Add preprint provider source tags when unregistered contributors are added to a preprint.
  - Add prereg/registered report campaign source tags when unregistered contributors are added to a node with prereg/registered report source tags.
  - Add OSF4M campaign source tag when unregistered contributors are added to a node with OSF4m campaign source tag.
  - Otherwise, add osf provider source tags.
- Add claimed tags to unregistered contributors when they claim their account.
  - Add preprint provider claimed tags when they claim from a preprint.
  - Add prereg/registered report campaign claimed tags when they claim from a node with corresponding source tags.
  - Add OSF4M campaign claimed tag when they claim from a node with the corresponding source tag.
  - Otherwise, add osf provider claimed tag.


## QA Notes

There's probably no good way to test the backfilling part of the ticket.
To test that new users are tagged properly, the following are the things to look for:
- Check that newly registered users and newly added unregistered contributors have proper source tags
  - If they register from osf home page, or they are added as an unreg contrib to a regular OSF node, they should have `source:provider|osf`.
  - If they register from preprint provider landing pages, or they are added as a unreg contrib to a preprint,they should have `source:provider|preprint|<provider_id>`
  - If they register from campaign landing pages, or they are added as an unreg contrib to a campaign node, they should have `source:campaign|<campaign_name>`.

- Check that confirmed unregistered contributors have claimed tags that corresponds to the node/preprint from which they claim their accounts.

## Side Effect

To normalize and backfill the tags, we need to run the two newly added scripts. The normalization script takes no time to run. However, the backfill scripts takes ~1 hour to run. (Time measured on a local instance of OSF with prod data.)

## Documentation

[Specs](https://docs.google.com/document/d/1vwnbPDAHrSOOSWydsSXGWO33QgOXf8NDdmZsWgmGPuc/edit#heading=h.13dsofnworfu)

[Existing tags and their relationship to source and claim tags](https://docs.google.com/spreadsheets/d/1C3idFKMNH8aOs22nv7P8QkA0zBujYJzA02VoEMNJDyU/edit#gid=0)

## Ticket

https://openscience.atlassian.net/browse/ENG-152

## Deployment Notes

Two scripts are to be run in the following order and requires system downtime:
`python -m scripts.normalize_user_tags`
`python -m scripts.backfill_source_tags_for_unregistered_contributors`
